README.md
# kg-disinfo

Knowledge Graph-based (dis)credibility estimation.

Given a KG and a few seed nodes which *contain (lack of) credibility scores* (represented as a numerical value), 
the system uses a metric propagation algorithm to estimate the (lack of) credibility of neighbouring nodes (for which no previous score is available). 

The algorithm allows you to define weights based on relation types: this allows 
you to define which relations should propagate more or less of the disinformation metric.

## Installation

Check the [releases](https://github.com/rdenaux/kg-disinfo/releases) and download the standalone jar.
Otherwise, you need to build it using leiningen as discussed below.

## Set up
The project is built using [Leiningen](https://leiningen.org/), which is like maven, but for [Clojure](https://clojure.org/).

## Usage

### With standalone jar
If you already have a standalone jar for this project you can call

    $ java -jar kg-disinfo-0.2.0-standalone.jar [args]
    
e.g. 

    $ java -jar kg-disinfo-0.2.0-standalone.jar --help 
    
will print the command line arguments. Similarly

    $ java -jar kg-disinfo-0.2.0-standalone.jar --graph resources/dante-g-small.json --injections resources/dante-g-small-injection.json
    
will:
  * read a graph from the specified path
  * read an `injection` from the specified path 
  * execute the `appleseed` algorithm on the graph and the seed injection, using  `0.9` as the `spreading-factor`
  * write the resulting scored graph to `C:/tmp/dante-g_scored.pdf`
  
### Using leiningen
Either 1. build the standalone jar by calling
   
    $ lein uberjar
    
and follow instructions described above, or

2. Use the `run` option, e.g.:

    $ lein run -- --help 
    
This will do the same as running it with the standalone jar as described above.

### Via repl (or programammatically)
Please refer to the code for examples on how to do this.

## Specifying graphs and injections

### graphs
The easiest way to use `kg-disinfo` is to provide your graph in a json format. 
A (automatically generated) [example is provided](https://github.com/rdenaux/kg-disinfo/blob/master/resources/dante-g-small.json). 
The file must contain:
* a single javascript object with attribute `edges`, which must be a list of edge objects. 
* each `edge` object has fields:
** `src-node` *mandatory* a string identifier for the source node
** `tgt-node` *mandatory* a string identifier for the target node
** `meta` metadata for the edge with required fields:
*** `weight`: *mandatory* this is the propagation weight between the source and the target node and should be a float between 0.0 and 1.0. Values close to 0.0 mean that (lack of) credibility scores from the source node will not be propagated to target node. Values close to 1.0 mean that (lack of) credibility scores from the source node will be propagated to the target node.
*** `name`: *optional* the name of the relation

Note that the graphs are directed, therefore, you may need to add inverse edges if appropriate. This is needed because (mis)credibility may propagate in one direction but not in the inverse.

Here's a short sample of what graph edges should look like.

```json
{
  "edges" : [ {
    "src-node" : "digitalId_23585",
    "tgt-node" : "person_23578",
    "meta" : {
      "weight" : 1.0,
      "name" : "isOwner-1"
    }
  }, {
    "src-node" : "digitalId_23585",
    "tgt-node" : "resource_23592",
    "meta" : {
      "weight" : 0.9,
      "name" : "isAuthor"
    }
  } ]
}
```

### injections
In order to propagate (mis)credibility scores, you first need to tell us what are your initial ratings. This is done by the `injections` file. An [example is provided](https://github.com/rdenaux/kg-disinfo/blob/master/resources/dante-g-small-injection.json):

```json
{
  "injection" : {
    "resource_23586" : 1.0,
    "resource_23594" : 1.0
  }
}
```

This is a json file with a single object with property `injection`. The `injection` should be another object with as properties, the node identifiers used in the corresponding `graph`. The values are the initial (mis)credibility scores, they *must* be float values. They should be larger than 0.0. We recommend using values between 0.0 and 1.0. Ideally these will have been assigned by human raters.

## Current limitations and improvement ideas

 * It would be nice to extend the `appleseed` algorithm to accept a `spreading-factor` value that depends on the node type. This would allow us to control that some intermediate nodes should not keep (much of the) score, while other nodes will tend to *keep* most of the scores.
 * When using the CLI, the client needs to specify the propagation weights when providing the graph. We could expose a way to assign default weights based on relationship types, but this would require providing additional information as part of the `graph`. Function `kg-disinfo.generator.generate-test-graph` already implements similar functionality which could be exposed and used for this. Similarly, the ability to automatically add inverse-edges could be specified as part of the graph metadata.

## Bugs

 * Not thoroughly tested

## Acknowledgements

Originally developed as part of the [DANTE](https://www.h2020-dante.eu/) project and refined as part of the [Co-inform](https://coinform.eu) project.

## License
Eclipse Public License 1.0
Copyright © 2019 Expert System Iberia


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Added
- Nothing

### Changed
- Nothing

### Removed
- Nothing

### Fixed
- Nothing


## [0.2.0] - 2019-11-07
### Added
- Defined json interface for providing input graphs and injections
- `kg-disinfo/core.main` now provides an CLI for executing injections over a graph

### Changed
- Refactored code to be more modular
- Improved README documentation

### Removed
- Nothing

### Fixed
- Nothing

## [0.1.0] - 2018-10-19
### Added
- Implementation of appleseed
- `dante-graphs` namespace to generate graphs according to the Dante operational ontology. This is described in D9.3 of Dante.

### Changed
- Nothing yet

### Removed
- Nothing

### Fixed
- No bugs yet

[Unreleased]: https://github.com/rdenaux/kg-disinfo/compare/0.2.0...HEAD
[0.2.0]: htpts://github.com/rdenaux/kg-disinfo/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/rdenaux/kg-disinfo/tags/0.1.0
	


test/kg_disinfo/core_test.clj
(ns kg-disinfo.core-test
  (:require [clojure.test :refer :all]
            [clojure.spec.alpha :as spec]
            [danteg]
            [kg-disinfo.core :refer :all]
            [kg-disinfo.appleseed :as appleseed]
            [kg-disinfo.dante-graphs :as dante-graphs]
            [ubergraph.core :as uber]))

(def readable-cfg
  ;; config useful for having small node objects (easier to inspect graph results)
  {:ent-map-fn (fn [ent] (-> ent :danteg/convertibleToCard :danteg/uuid))
            :rel-map-fn (fn [rel] (-> rel :danteg/relation :danteg/relationUuid))}
  )

(comment 
  (deftest test-parse-json-dante-graph
    (let [json-str (slurp "resources/graph_for_disinformation_1.txt")
          dante-g (dante-graphs/parse-json-dante-graph json-str)]
      (when (not (spec/valid? :danteg/dante-graph dante-g))
        (do 
                                        ;(spec/explain :danteg/dante-graph dante-g)
          (is false "does not conform to spec"))
        ))))

(deftest test-write-json-dante-graph
  (let [dante-g {:danteg/id 5 :danteg/uuid "asdf"}]
    (is (= (dante-graphs/write-json-dante-graph dante-g)
           "{\"id\":5,\"uuid\":\"asdf\"}")))
  )


(deftest test-read-seed-injections
  (let [json-str (slurp "resources/graph_for_disinformation_1.txt")
        dante-g (dante-graphs/parse-json-dante-graph json-str)
        ;g (dante-graphs/create-from-dante-graph dante-g readable-cfg)
        seed-injections (dante-graphs/read-seed-injections dante-g readable-cfg)]
    (println "seed injections " seed-injections)
    (is (not (empty? seed-injections)))
    ))

(defn =roundf
  [dec-points a b]
  (assert (int? dec-points))
  (assert (> dec-points 0))
  (let [format-exp (str "%." dec-points "f")]
    (= (format format-exp a) (format format-exp b))))

(deftest test-appleseed-on-dante-test-graph_1
  (let [json-str (slurp "resources/graph_for_disinformation_1.txt")
        dante-g (dante-graphs/parse-json-dante-graph json-str)
        g (dante-graphs/create-from-dante-graph dante-g readable-cfg)
        seed-injections (dante-graphs/read-seed-injections dante-g readable-cfg)
        spreading-factor 0.9
        acc-threshold 0.001]
    (println "dante-g entities type " (type (:danteg/uiEntitiesList dante-g)))
    (let [prop-scores (time (appleseed/appleseed* g seed-injections spreading-factor acc-threshold))]
      (is (not (empty? prop-scores)))
      (println "prop scores for test graph 1: " prop-scores)
      (is (=roundf 5 0.725538 (get prop-scores "KB-KB_API-Person_08455784-9011-4dbf-b506-1df98cfa2b02")))
      (is (=roundf 5 0.178389 (get prop-scores "KB-KB_API-Event_e186935a-d199-4654-b80d-130d9eb8df7e")))
      (is (=roundf 5 0.334531 (get prop-scores "KB-KB_API-Person_253ed624-1d20-44ee-a1a2-1c160561da9f")))
      (is (=roundf 5 0.405504 (get prop-scores "KB-KB_API-Organization_148cca61-d841-457e-b73f-c93f0f7d2405")))
      (is (=roundf 5 0.334531 (get prop-scores "KB-KB_API-Person_9527738b-d61b-4af3-be03-116cfe5cb387")))
      )
    ))

(deftest test-appleseed-on-dante-test-graph_2
  (let [json-str (slurp "resources/graph_for_disinformation_2.txt")
        dante-g (dante-graphs/parse-json-dante-graph json-str)
        g (dante-graphs/create-from-dante-graph dante-g readable-cfg)
        seed-injections (dante-graphs/read-seed-injections dante-g readable-cfg)
        spreading-factor 0.9
        acc-threshold 0.001]
    (let [prop-scores (time (appleseed/appleseed* g seed-injections spreading-factor acc-threshold))]
      (is (not (empty? prop-scores)))
      (is (= (count seed-injections) (count prop-scores)))
      )
    )
  )

(deftest test-read-danteg
  (let [g (dante-graphs/create-from-dante-graph
           "resources/graph_for_disinformation_1.txt" readable-cfg
           )]
    (uber/pprint g)
    (is (not (uber/undirected-graph? g)))
    (is (= 5 (count (uber/nodes g))))
    (is (= 4 (count (uber/edges g))))
  ))

(deftest test-process-dante-graph-json_1
  (let [json-str (slurp "resources/graph_for_disinformation_1.txt")
        out-json (process-dante-graph-json json-str)]
    (spit "target/in-graph-1.json" json-str)
    (spit "target/out-graph-1.json" out-json)
    (is (not (= json-str out-json)))
    ))

(comment 
  (deftest a-test
    (testing "FIXME, I fail to generate graph?"
      (let [g-spec {:rel-counts {:isOwner 2 :isAuthor 2}}
            g (dante-graphs/generate-dante-test-graph g-spec)]
        (uber/pprint g)
        (is (= 0 1)))
      )))


project.clj
(require 'cemerick.pomegranate.aether)
(cemerick.pomegranate.aether/register-wagon-factory!
  "http" #(org.apache.maven.wagon.providers.http.HttpWagon.))
(defproject net.expertsystem.lab/kg-disinfo "0.2.0"
  :description "Disinformation estimation through spreading activation"
  :url "https://github.com/rdenaux/kg-disinfo"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.9.0"]
                 [org.clojure/tools.logging "0.4.1"]
                 [ch.qos.logback/logback-classic "1.1.3"]
                 [org.clojure/tools.cli "0.4.2"]
                 [cheshire "5.8.1"]
                 [ubergraph "0.5.1"]]
  :plugins [[refactor-nrepl "2.4.0"]
           [cider/cider-nrepl "0.17.0"]]
  :deploy-repositories [["esi-private" {:url "http://nexus.iberia.expertsystem.local/repository/esi-private"
                                        :sign-releases false}]]
  :jvm-opts ["-Dfile.encoding=UTF-8"]
  :main ^:skip-aot kg-disinfo.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/danteg.clj
(ns danteg
  "Defines a spec for the Dante graphs"
  (:require [clojure.spec.alpha :as s]))

(s/def ::targetEntity string?)
(s/def ::pid string?) ;;UUID?
(s/def ::date string?)
(s/def ::id int?)

(s/def ::typed
  (s/keys :req-un [::type]))

(s/def ::disinformationScore float?)

(s/def ::person
  (s/keys :req-un [::type
                   ]))
(s/def ::resource
  (s/or :video ::video
        :image ::image
        :text  ::text))

(s/def ::convertibleToCard
  (s/and 
   (s/keys :req-un [::type])
   (s/or :person ::person
         :event  ::event
         :org    ::organization
         :resource ::resoure
         :digitalId ::digitalIdentity
         )))

(s/def ::uiEntity
  (s/keys :req-un [::id ::convertibleToCard ::disinformationScore]))

(s/def ::uiEntitiesList (s/+ ::uiEntity))

(s/def ::relationUuid string?)
(s/def ::relationType (= "relation"))
(s/def ::relationCategory
  (s/and string?
         (s/or
          :isBrother   #(= % "#isBrother") ;; person 2 person
          :isOwner     #(= % "#isOwner") ;; person 2 digid
          :isInvolved  #(= % "#isInvolved"); digid 2 event
          :isAuthor    #(= % "#isAuthor"); (or video text) 2 digid
          :follow      #(= % "#follow"); digid 2 digid 
          :like        #(= % "#like"); digid 2 digid
               )))
(s/def ::startNodeUUid string?)
(s/def ::endNodeUuid string?)

(s/def ::relation
  (s/keys :req-un [::relationUuid ::relationType
                   ::relationCategory
                   ::startNodeUUid
                   ::endNodeUuid])
  )
(s/def ::uiRelation
  (s/keys :req-un [::source ::target ::relation]))

(s/def ::uiRelationList (s/+ ::uiRelation))

(s/def ::dante-graph
  (s/keys :req-un [::targetEntity ::pid
                   ::date
                   ::uiEntititesList
                   ::uiRelationList]))



src/kg_disinfo/io.clj
(ns kg-disinfo.io
  "Read and write graphs to files and supported formats"
  (:require [clojure.java.io :as io]
            [ubergraph.core :as uber]
            [cheshire.core :as cheshire]
            [clojure.string :as str]
            [clojure.tools.logging :as log]))

(defn strip-ns
  [keyword]
  (let [result (if (qualified-keyword? keyword)
                 (keyword (str/replace (name keyword) "jsonkg/" ""))
                 keyword)]
    (log/debug (str "Stripped " keyword " = " result))
    result))

(defn json->uber-edge
  [json-edge]
  [(:jsonkg/src-node json-edge)
   (:jsonkg/tgt-node json-edge)
   (->>  (:jsonkg/meta json-edge)
        (map (fn [[k v]]
               [(keyword (name k)) v]))
        (into {}))])

(defn jsonKG->edges
  "Converts a `:jsonKG/graph` datastructure into a vector of uber edges"
  [json-map]
  (log/debug (str  "Reading with map with keys " (keys json-map)
               " and " (count (:jsonkg/edges json-map)) " edges"))
  (->> (:jsonkg/edges json-map)
       (map json->uber-edge)
       (into [])))

(defn parse-json-kg
  [json-str]
  (cheshire/parse-string json-str (fn [key] (keyword (str "jsonkg/" key)))))

(defn write-json-kg
  [jsonKG opts]
  (cheshire/generate-string
   jsonKG
   (merge opts {:key-fn
                ;;(fn [key] (str/replace key ":jsonkg/" ""))
                (fn [key] (name key))
                })))

(defn read-json-digraph
  [path-or-json-str]
  (let [json-str (if (.exists (io/as-file path-or-json-str))
                   (slurp path-or-json-str)
                   path-or-json-str)
        json-map (parse-json-kg json-str)
        edges (jsonKG->edges json-map)]
    (log/info (str "Creating digraph from " (count edges) " edges"))
    (log/info (str "Example edge " (first edges)))
    (apply uber/digraph edges))
  )

(defn uber->json-edge
  [uber-g uber-edge]
  #:jsonkg{:src-node (:src uber-edge)
           :tgt-node (:dest uber-edge)
           :meta (uber/attrs uber-g uber-edge)})

(defn uberg->jsonkg
  [uber-g]
  {:jsonkg/edges (->> (uber/edges uber-g)
                      (map (partial uber->json-edge uber-g))
                      (into [])
                     )})

(defn write-as-json-kg
  [uber-g out-path json-gen-opts]
  (spit out-path
        (-> uber-g
            (uberg->jsonkg)
            (write-json-kg (or json-gen-opts {})))
        ))


;;;
;;;  injections (maps from node ids to initial values to propagate
;;;

(defn parse-json-injection
  [json-str]
  (cheshire/parse-string json-str))

(defn read-injection-json
  [path-or-json-str]
  (let [json-str (if (.exists (io/as-file path-or-json-str))
                   (do
                     (log/info (str "Reading injections from: " path-or-json-str))
                     (slurp path-or-json-str))
                   path-or-json-str)
        inj-map (parse-json-injection json-str)
        json-map (get inj-map "injection")
        ]
    (log/info (str "Read injection with " (count json-map) " injections"))
    json-map))

(defn write-json
  [value opts]
  (cheshire/generate-string
   value
   (merge opts {:key-fn (fn [key] (name key))
                })))

(defn write-injection-json
  [injection-map out-path json-gen-opts]
  (spit out-path
        (write-json {:injection injection-map} (or json-gen-opts {})))
  )

;;;
;;;
;;;
(defn write-propagated-scores-json
  [scores out-path json-gen-opts]
  (spit out-path
        (write-json {:scores scores} (or json-gen-opts {}))))


src/kg_disinfo/appleseed.clj
(ns kg-disinfo.appleseed
  "Implements the appleseed algorithm"
  (:require [ubergraph.core :as uber]
            ))

(defn new-nodes-in
  "Returns a set of new target-nodes that are reachable from source-nodes"
  [graph source-nodes]
  (let [tgt-nodes (mapcat
                   (fn [src-node] ;; get-target-nodes from src-node
                     (map :dest (uber/find-edges graph {:src src-node})))
                   source-nodes)]
    (into #{} (filter (fn [node] (not (contains? source-nodes node))) tgt-nodes))
    ))

(defn trust-change
  [spreading-factor injection-map]
  (zipmap (keys injection-map)  
          (map (fn [k] (* (- 1 spreading-factor)
                         (get injection-map k)))
               (keys injection-map))))

(defn combine-maps
  [map-a map-b val-merge-fn defval-a defval-b]
  (let [keys-ab (into (into #{} (keys map-a)) (keys map-b))]
    (zipmap keys-ab
            (map (fn [k] (apply val-merge-fn [(get map-a k defval-a) (get map-b k defval-b)]))
                 keys-ab)
            )))


(defn calc-in-contribs
  "Returns a LazySeq with the weighted contributions for each in-edge"
  [in-edges in_prev graph]
  (map (fn [in-edge]
         (let [src-node (:src in-edge)
               edge-weight (uber/attr graph in-edge :weight)
               out-edges (uber/find-edges graph {:src src-node})
               out-weights (map (fn [e] (uber/attr graph e :weight)) out-edges)]
           (* (get in_prev src-node 0.0) ;; use 0.0 if no previous value known
              (/ edge-weight
                 ;;avoid division by 0
                 (+ 0.0001 (apply + out-weights)))))
         )
       in-edges)
  )

(defn update-incoming
  "Returns an updated version of in_prev
  for the nodes already in in_prev and the new nodes"
  [in_prev graph spreading-factor V_new]
  (let [nodes (into (keys in_prev) V_new)]
    (zipmap nodes
            (map (fn [k]
                   (let [in-edges (uber/find-edges graph {:dest k})
                         in-contribs (calc-in-contribs in-edges in_prev graph)]
                     ;;(print "weighted input contributions for node " k ": " in-contribs "\n")
                     (* spreading-factor
                        (apply + in-contribs))
                     ))
                 nodes)))
  )

(defn add-virtual-nodes
  [graph seed-node new_nodes]
  (uber/add-directed-edges*
   graph
   (map (fn [node]
          [node seed-node {:weight 1.0 :name :appleseed-virtual}])
        new_nodes))
  )

(defn appleseed
  "Implements the appleseed trust metric.
  Returns a map from nodes in the graph to updated scores.
  see also https://pdfs.semanticscholar.org/2635/19ea851c96e67b550bf03b4ec8922beeb120.pdf
  "
  [graph seed-node trust-injection spreading-factor accuracy-threshold]
  (loop [i 0
         g_i graph
         V_i #{seed-node}
         trust_i {seed-node 0.0}
         in_i {seed-node trust-injection}]
    (let [V_new (new-nodes-in g_i V_i)
          trust-diff (trust-change spreading-factor in_i) 
          trust_new (into (combine-maps trust_i trust-diff + 0.0 0.0)
                          (zipmap V_new (repeat (count V_new) 0.0)))
          max-change (apply max (vals trust-diff))
          g_new (add-virtual-nodes g_i seed-node V_new)]
      ;;(print "appleseed it" i "\n")
      (comment 
        (print "\tnew nodes: " V_new
               "\n\tmax-trust-change " max-change
               "\n\tinjection" in_i
               "\n\ttotal-trust assigned" (apply + (vals trust_new))
               "\n"))
      (if (and (<= max-change accuracy-threshold)
               (empty? V_new))
        trust_new
        (recur (inc i)
               g_new
               (into V_i V_new)
               trust_new
               (update-incoming in_i g_i spreading-factor V_new)
                     ))
      )))

(defn appleseed*
  "Apply more than one trust injection into a graph using appleseed.
  `graph` must be an `uber/digraph`
  `trust-injections` is a map from nodes to the seed score
  `spreading-factory` is a float between 0.0 and 1.0
  `accuracy-threshold` is a float larger than but close to 0.0. The bigger
    it is, the sooner each iteration will stop.
  "
  [graph trust-injections spreading-factor accuracy-threshold]
  (reduce (fn [acc  trust-map]
            (combine-maps acc trust-map + 0.0 0.0))
          (map (fn [seed-node]
                 (appleseed graph seed-node (get trust-injections seed-node)
                            spreading-factor accuracy-threshold))
               (keys trust-injections)))
  )


src/kg_disinfo/generator.clj
(ns kg-disinfo.generator
  "Provides methods for generating a random graph based on a specification
  ```clj
    (generator/generate-test-graph
     {
       :nodes {:person 3 :article 6 :publisher 2}
       :rel-counts {:authorOf 2 :publishedOn 3 :follows 4}
       :inverse-rel-map {:authorOf :byAuthor
                         :publishedOn :published}
       :default-reltype-weights
       {:authorOf {:source :article :target :person :weight 1.0 :comment \"credible articles should give the author credibility\"}
        :byAuthor {:source :person :target article :weight 1.0 :comment \"credible authors write credible articles\")
        :publishedOn {:source :article :target :publisher :weight 1.0}
        :published {:source :publisher :target :article :weight 0.5}
        :follows {:source person :target person :weight 0.1 :comment \"credible people don't necessarily follow credible people\"}}})
  ```"
  (:require [ubergraph.core :as uber]
            ))

;;;
;;; Node generation
;;;

(defn generate-symbols
  "Generates `n` unique symbols starting with the specified prefix"
  [n prefix]
  (map (fn [ignore] (str (gensym prefix))) (range n)))

(defn generate-nodes
  [gen-spec]
  (let [nodespec (:nodes gen-spec)
        gen-symbs (fn [node-type]
                    (generate-symbols (node-type nodespec 3)
                                   (str (name node-type) "_")))]
    (->> (keys nodespec)
         (map (fn [node-type]
                [node-type (gen-symbs node-type)]))
         (into {}))))


;;;
;;; Edge generation
;;;

(defn get-nodes
  [source nodes-by-type]
  (cond
    (keyword? source) (source nodes-by-type)
    (set? source) (get-nodes (rand-nth (into [] source)) nodes-by-type)
    :else (let [msg (str "Unknown source type " source "?\n")]
           (print msg)
           [])
    ))

(defn select-node-pair
  [{:keys [source target]} nodes-by-type]
  (let [source-candidates (get-nodes source nodes-by-type)
        target-candidates (get-nodes target nodes-by-type)]
    (if (nil? source-candidates)
      (print "nodes-by-type does not contain " source "?"
             (keys nodes-by-type) "?\n"))
    (if (nil? target-candidates)
      (print "nodes-by-type does not contain " target "?"
             (keys nodes-by-type) "?\n"))
    [(rand-nth source-candidates)
     (rand-nth target-candidates)]
  ))

(defn generate-edge
  [cfg rel-type nodes-by-type]
  (let [rel-meta (get-in cfg [:default-reltype-weights rel-type])
        pair (select-node-pair rel-meta nodes-by-type)
        ]
    ;;(print "edge for random pair" pair "\n")
    (conj pair {:weight (:weight rel-meta) :name (name rel-type)})
    ))

(defn generate-n-edges
  [cfg [rel-type count] nodes-by-type]
  (map (fn [ignore] (generate-edge cfg rel-type nodes-by-type)) (range count)))

(defn generate-edges
  "Generates relations based on a generation spec and a map of nodes by type"
  [gen-spec nodes-by-type]
  (mapcat (fn [rel-type-count] (generate-n-edges gen-spec rel-type-count nodes-by-type))
          (seq (:rel-counts gen-spec))))


;;;
;;; Inverse edge generation
;;;

(defn invertible-edge?
  [[src-node tgt-node attrs] cfg]
  (contains? (into #{} (keys (:inverse-rel-map cfg))) (keyword (:name attrs))))

(defn generate-inverse
  [edges cfg]
  (->> edges
       (filter (fn [edge] (invertible-edge? edge cfg)))
       (map (fn [[src-node tgt-node attrs]]
              (let [rel-type (keyword (:name attrs))
                    inv-type (get-in cfg [:inverse-rel-map rel-type]) 
                    inv-weight (get-in cfg
                                       [:default-reltype-weights inv-type :weight]) 
                    ]
                [tgt-node src-node {:weight inv-weight :name (name inv-type)}]))
            )))

(defn add-inverse-edges
  [edges cfg]
  (into edges (generate-inverse edges cfg)))


;;;
;;;  Main graph generation function 
;;;

(defn generate-test-graph
  "Generates a digraph based on a generation spec
  The `gen-spec` must be a map with keys `:nodes`, `:rel-counts`,
  `:inverse-rel-map` and `:default-reltype-weights`"
  [gen-spec]
  (let [V (generate-nodes gen-spec)
        E (generate-edges gen-spec V)
        cfg gen-spec]
    (apply uber/digraph (add-inverse-edges E cfg))))


src/kg_disinfo/core.clj
(ns kg-disinfo.core
  (:require [ubergraph.core :as uber]
            [kg-disinfo.dante-graphs :as dante-graphs]
            [kg-disinfo.io :as kgio]
            [kg-disinfo.generator :as kgen]
            [kg-disinfo.appleseed :as appleseed]
            [clojure.tools.cli :refer [parse-opts]]
            [clojure.string :as str]
            [clojure.java.io :as io]
            [clojure.tools.logging :as log])
  (:gen-class)
  )

(defn process-dante-graph-json
  "Accepts a dante-g-json string and produces a modified version
  with updated disinformation scores."
  [dante-g-json]
  (let [cfg {:ent-map-fn identity
             :rel-map-fn identity}
        dante-g (dante-graphs/parse-json-dante-graph dante-g-json)
        g (dante-graphs/create-from-dante-graph dante-g cfg)
        seed-injections (dante-graphs/read-seed-injections dante-g cfg)
        spreading-factor 0.9
        acc-threshold 0.001]
    (let [prop-scores (appleseed/appleseed* g seed-injections spreading-factor acc-threshold)]
      (dante-graphs/write-json-dante-graph (dante-graphs/update-dante-g dante-g prop-scores cfg))
      )))

(def cli-options
  [
   ["-g" "--graph FILE" "Path to a json file containing a graph description"
    :validate [#(.exists (io/file %)) "Graph file must exist"]]
   ["-i" "--injections FILE" "Path to a json file containing the initial scores"
    :validate [#(.exists (io/file %)) "Injections file must exist"]]
   ["-s" "--spreading-factor double" "Float value between 0 and 1"
    :default 0.85
    :parse-fn #(Double/parseDouble %)
    :validate [#(< 0.0 % 1.0) "Must be a number between (but excluding!) 0.0 and 1.0"]
    ]
   [nil "--accuracy-threshold double" "Float value between 0 and 0.1"
    :default 0.001
    :parse-fn #(Double/parseDouble %)
    :validate [#(< 0.0 % 0.1) "Must be a number betwee 0.0 and 0.1"]]
   [nil "--generate-viz format"
    "Format for generating the base, injected and scored graph visualizations. Options pdf, png, svg"
    :parse-fn #(keyword %)
    :validate [#(contains? #{:png :pdf :svg} %) "Must be either png, pdf or svg"]]
   ["-h" "--help"]
   ])

(defn- usage [options-summary prog-name]
  (->> [prog-name
        ""
        (str "Usage: " prog-name " [options]")
        ""
        "Options:"
        options-summary
        ""
        "Please refer to the documentation on github for more information."
        "Copyright 2018-2019 Expert System"]
       (clojure.string/join \newline)))

(defn- error-msg [errors]
  (str "The following errors occurred while parsing your command:\n\n\t"
       (str/join "\n\t" errors)))

(defn- cannot-execute-with-opts
  [options]
  (cond
    (not (:graph options)) "graph option is missing"
    (not (:injections options)) "injections option is missing"
    :default nil
    ))

(defn- validate-args
  "Validate command line arguments"
  [args prog-name cli-options]
  (let [{:keys [options arguments errors summary]} (parse-opts args cli-options)]
    (cond
       errors
       {:exit-message (error-msg errors)}
       
       (not errors)
       (do
         (when (:help options)
           (log/info (usage summary prog-name)))         
         (let [err-msg (cannot-execute-with-opts options)]
           {:options options
            :exit-message (when err-msg ;; do not repeat help 
                            (str (when-not (:help options)  (usage summary prog-name))
                                 "\n" err-msg))
            :ok? (not err-msg)}))
       
       :else
       {:exit-message (usage summary prog-name)
        :ok? true}))
    )

(defn- sys-exit [status msg]
  (if (= 0 status)
    (log/info msg)
    (log/error msg))
  (System/exit status))

(defn- run
  [opts]
  (log/info (str  "Executing kg-dinsinfo with options" opts))
  (let [graph-path (:graph opts)
        graph (kgio/read-json-digraph graph-path)
        seed-injections (kgio/read-injection-json (:injections opts))
        spreading-factor (:spreading-factor opts)
        accuracy-threshold (:accuracy-threshold opts)
        base-path (str/replace graph-path ".json" "") ;;(.getParent (io/as-file graph-path))
        gen-viz (:generate-viz opts)
        ]
    (assert (< 0 (count seed-injections))
            (str "Must have at least one injection"))
    (assert (not-empty (clojure.set/intersection
                        (set (keys seed-injections))
                        (set (uber/nodes graph))))
            (str  "None of the injections are nodes in the graph. "
                  "Verify injection and graph files correspond to each other."))
    (assert (< 0.0 spreading-factor 1.0))
    (assert (< 0.0 accuracy-threshold 0.1))
    (when gen-viz
      (log/info "Storing base graph viz at" (str base-path "_base." (name gen-viz)))
      (uber/viz-graph graph
                      {:save {:filename (str base-path "_base." (name gen-viz)) :format gen-viz}}))
    (when gen-viz
      (log/info "Storing base graph with injections\n")
      (uber/viz-graph (dante-graphs/color-by-scores graph seed-injections)
                      {:save {:filename (str base-path "_base_with_seed." (name gen-viz)) :format gen-viz}}))
    (let [propagated-scores (time (appleseed/appleseed*
                                   graph seed-injections spreading-factor accuracy-threshold))]
      (when gen-viz
        (log/info "Storing scored graph")
        (uber/viz-graph (dante-graphs/color-by-scores graph propagated-scores)
                        {:save {:filename (str base-path "_base_scored." (name gen-viz)) :format gen-viz}}))
      (log/info "Storing updated scores at " (str base-path "_propagated.json"))
      (kgio/write-propagated-scores-json propagated-scores
                                         (str base-path "_propagated.json")
                                         {:pretty true})
      (log/info "Finished")
    )))

(defn -main
  "Reads the graph and injections from files given by the args and executes the injection propagation.
  Outputs a json file with the output scores as well as files with graph visualizations if requested."
  [& args]
  (let [{:keys [options exit-message ok?]} (validate-args args "kg-disinfo" cli-options)]
    (if exit-message
      (sys-exit (if ok? 0 1)
                exit-message)
      (run options)
      ))
  (sys-exit 0 "done"))
